### PR TITLE
Protect against DOS Plugin Registration

### PR DESCRIPTION
**Pull request by mattjmcnaughton**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Use token bucket rate limiting to protect against denial of service
attacks (intentional or unintentional) when registering new/updatable plugins.

**Which issue(s) this PR fixes**:
Address TODO in the code base.

**Special notes for your reviewer**:
cc @taragu is that what you envisioned?

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by mattjmcnaughton** [comment]
@taragu would love your thoughts here :) is that what you envisioned? thanks!

**Comment by mattjmcnaughton** [comment]
@taragu would love your thoughts here :) is that what you envisioned? thanks!

**Comment by mattjmcnaughton** [comment]
@taragu would love your thoughts here :) is that what you envisioned? thanks!

**Comment by taragu** [comment]
Thank you for the PR @mattjmcnaughton! I’ll take a closer look at this this week. Also cc’ing @saad-ali here.
/cc @saad-ali

File: pkg/kubelet/pluginmanager/plugin_manager.go [hunk 1]
```diff
@@ -68,11 +69,16 @@ func NewPluginManager(
 		asw,
 	)
 
+	// We rate limit registering new plugins to prevent denial of service.
+	var pluginRegistrationQPS float32 = 10.0
```
Comment (taragu) [inline pkg/kubelet/pluginmanager/plugin_manager.go]: These should be consts instead of vars. Also I'm not sure if these are what to want to set in the code, or users should be able to set this in the config 🤔 @saad-ali what's your opinion on this?

**Review by taragu — COMMENTED** [review]
Can we also have some test cases for exceeding/passing registration QPS, with a fake `RateLimiter`?

**Comment by mattjmcnaughton** [comment]
/hold

I haven't forgotten about this diff :) I just need to find the time to write the requested unit tests. Marking it as hold so people know not to review :)

**Comment by mattjmcnaughton** [comment]
/remove hold

@taragu I added unit tests! Sorry for the delay :)

File: pkg/kubelet/pluginmanager/pluginwatcher/plugin_watcher.go [hunk 2]
```diff
@@ -17,35 +17,45 @@ limitations under the License.
...
 	"github.com/fsnotify/fsnotify"
+	"k8s.io/client-go/util/flowcontrol"
 	"k8s.io/klog"
 
 	"k8s.io/kubernetes/pkg/kubelet/pluginmanager/cache"
 	"k8s.io/kubernetes/pkg/kubelet/util"
 	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
 )
 
+const (
```
Comment (mattjmcnaughton) [inline pkg/kubelet/pluginmanager/pluginwatcher/plugin_watcher.go]: I kind of just made these values up :) In my mind, they don't need to be super fine tuned ... we're mostly just focusing on preventing clear DOS situations. But definitely happy to hear any feedback.

File: pkg/kubelet/pluginmanager/pluginwatcher/plugin_watcher_test.go [hunk 3]
```diff
@@ -253,8 +254,55 @@ func TestPluginRegistrationAtKubeletStart(t *testing.T) {
...
+	// for the number we attempt to create.
+	numPluginsToAttemptCreate := 1
+	for i := 0; i < numPluginsToAttemptCreate; i++ {
+		socketPath := fmt.Sprintf("%s/plugin-%d.sock", socketDir, i)
+		pluginName := fmt.Sprintf("example-plugin-%d", i)
+
+		p := NewTestExamplePlugin(pluginName, registerapi.DevicePlugin, socketPath, supportedVersions...)
+		require.NoError(t, p.Serve("v1beta1", "v1beta2"))
+
+		// Ideally, we'd be able to wait and see the
```
Comment (mattjmcnaughton) [inline pkg/kubelet/pluginmanager/pluginwatcher/plugin_watcher_test.go]: Particularly curious to get feedback on this component.

**Comment by mattjmcnaughton** [comment]
/assign @saad-ali

@saad-ali , when you get a chance, do you mind taking a look at this diff to address a TODO around testing aspects of plugin registration? thanks!

**Review by saad-ali — COMMENTED** [review]
What is the DOS scenario we are trying to prevent here?  Is it a malicious user or malfunctioning plugin repeatedly creating/deleting a socket file? How realistic is that? More importantly, what are we trying to protect? Trying to understand if introducing a global rate limiter that adds a wait helps here.

**Review by mattjmcnaughton — COMMENTED** [review]
> What is the DOS scenario we are trying to prevent here? Is it a malicious user or malfunctioning plugin repeatedly creating/deleting a socket file? How realistic is that? More importantly, what are we trying to protect? Trying to understand if introducing a global rate limiter that adds a wait helps here.

cc @taragu who is the original author of the TODO :)

I would imagine we are more concerned with preventing a DOS from a malfunctioning plugin which is attempting to register/deregister itself repeatedly. My guess is that if our cluster has been comprised such that a malicious actor is able to add/remove plugins at will, we have bigger issues to worry about...

W.r.t. what will happen if the "denial of service" occurs, if I'm reading the code correctly, the `DesiredStateOfWorld` would be updated continuously as this plugin is added/removed. This operation just involves claiming a lock and then updating a hash. The reconciler already is "rate limited" in that it runs once a second. So tl:dr; I don't think there are any serious ramifications of a DOS, given the way the code is written. The only possible issues I could foresee would be the cost of continuously claiming/releasing locks and potentially log spam.

As to whether this failure scenario prevents an _actual_ risk is a very fair question... my guess is that the odds of this happening are fairly small. I certainly haven't heard any reports of it happening in clusters "in the wild". Additionally, at least in my grepping of the code base, I don't see a ton of other rate limiters used to prevent DOS (i.e. I don't think its a standard pattern in k8s).

On the other hand... if I'm understanding how the rate limiting works, this new code should have no impact on behavior unless we are experiencing heavy load (in which case it will rate limit). So I think the only real cost here is the slightly higher complexity of the additional code.

tl:dr; I'm happy with either moving forward with this code or creating a PR to remove the TODO. I don't think we should leave the code as is, as it creates the risk of another dev attempting to address the TODO and us having this same convo again :)

Thanks @saad-ali !

**Comment by mattjmcnaughton** [comment]
@saad-ali I filed https://github.com/kubernetes/kubernetes/pull/88266 which just deletes the TODO. If you don't mind accepting either this PR or that PR, we can wrap this thread up. Thanks for your time :)

**Comment by saad-ali** [comment]
> tl:dr; I'm happy with either moving forward with this code or creating a PR to remove the TODO. I don't think we should leave the code as is, as it creates the risk of another dev attempting to address the TODO and us having this same convo again :)

> @saad-ali I filed #88266 which just deletes the TODO. If you don't mind accepting either this PR or that PR, we can wrap this thread up. Thanks for your time :)

Agreed, and thank you for your detailed response and all your hard work on this. I'd prefer to keep the code simple unless we have good reason to do something more complicated, so let's close this PR and go with #88266

**Comment by mattjmcnaughton** [comment]
> > tl:dr; I'm happy with either moving forward with this code or creating a PR to remove the TODO. I don't think we should leave the code as is, as it creates the risk of another dev attempting to address the TODO and us having this same convo again :)
> 
> > @saad-ali I filed #88266 which just deletes the TODO. If you don't mind accepting either this PR or that PR, we can wrap this thread up. Thanks for your time :)
> 
> Agreed, and thank you for your detailed response and all your hard work on this. I'd prefer to keep the code simple unless we have good reason to do something more complicated, so let's close this PR and go with #88266

Sounds like a plan! Closing this PR in favor of https://github.com/kubernetes/kubernetes/pull/88266.
